Articles/Blog/TeamExperiencesImplementingCI.md

```diff
@@ -0,0 +1,82 @@
+
+Like VisIt, ZFP doesn’t require 100% of its tests to pass before merging unless the merge is going into the main development branch. This approach allows the main branch to serve as a base for releases while letting the feature branches be a place to work collaboratively on issues.
+
+### Summary
+
+As a technology, CI has evolved and developed considerably over the past several years.  Mature systems, connected to ubiquitous version control platforms like GitHub, GitLab, and Bitbucket, offer many examples, tutorials, and documentation that make the entry barrier low. The projects interviewed in this article discussed their specific adaptations for scientific and numerical software.  All three projects mentioned comparing with “golden results” from full-program runs, as well as the unique mindset needed to develop and maintain unit tests.  Testing is a mode and scope of work motivated by adding armor to your codebase. The benefit most cited here was identifying potential bugs early, thereby increasing the project’s ability to collaborate within the open source community.
+
+All three projects also noted some drawbacks to CI: maintaining tests requires time and effort. Running these tools requires trial-and-error, and long-running tests can become a source of frustration. Teams noted that random failures (due to network, and other sources) were initially annoying but infrequent enough to safely ignore and usually resolved by a restart. Long-running tests can stress CI resources, with the mixed benefit/drawback of needing more careful attention to how tests are designed.  Other than adding that development work, the teams didn’t notice excessively disruptive changes to their process.
+
+Aside from CI, much can be learned from studying code construction choices across open-source projects.  For example, the three projects here all use a different strategy to create Python interfaces to their codes.  ExaSMR uses Python’s ctypes to wrap the C API, then installs those wrappers in site-packages so they can be found by setting PYTHONPATH.  VisIt has an integrated Python environment.  ZFP’s Python interface is built using [Cython’s support for wrapping C++ classes](https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#using-c-in-cython), with a build and install process managed using the [scikit-build ](http://www.scikit-build.org) plugin to CMake.
+
+Overall, CI adds value by providing more stable working code bases, test feedback without developer intervention, and greater support for community contributions.  As we watch the evolution of CI technologies, our teams are seeking several features. The ability to freeze the state of tests would both reduce redundant tests after small changes and also allow interaction with these states. Greater availability of testing hardware and runners would further boost the usefulness of this technology.
+
+### Acknowledgements
```

Comment:
@frobnitzem would it be appropriate to thank your primary contacts in each team individually?  As mentioned above, I think this would be a more appropriate place for the links to go if you really want them.

Reply:
- [ ]